### PR TITLE
types: Use `IPAddress` instead of `str` in `PeerInfo`

### PR DESCRIPTION
**Pull request by xdustinface**
Working towards using `IPAddress` instead of `str` all over / where it makes sense.

Based on:

- #14166

### CODE REVIEW

File: chia/server/address_manager.py [hunk 1]
```diff
@@ -300,7 +300,7 @@ def clear_new_(self, bucket: int, pos: int) -> None:
     def mark_good_(self, addr: PeerInfo, test_before_evict: bool, timestamp: int) -> None:
         self.last_good = timestamp
         (info, node_id) = self.find_(addr)
-        if not addr.is_valid(self.allow_private_subnets):
+        if addr.ip.is_private and not self.allow_private_subnets:
```
Comment (arvidn) [inline chia/server/address_manager.py:303]: it's not obvious that the new condition is the same as the old. Is being in a private subnet while private subnets are disallowed the only way an IP can be invalid in?
The left side suggests the IP address can be invalid in more ways

File: chia/server/node_discovery.py [hunk 2]
```diff
@@ -62,7 +63,7 @@ def __init__(
         random.shuffle(dns_servers)  # Don't always start with the same DNS server
         if introducer_info is not None:
             self.introducer_info: Optional[PeerInfo] = PeerInfo(
-                introducer_info["host"],
+                str(get_host_addr(introducer_info["host"], prefer_ipv6=False)),
```
Comment (arvidn) [inline chia/server/node_discovery.py:66]: is it OK to make a blocking call here? I assume `get_host_addr()` is blocking in case it needs to perform a DNS lookup. If it's OK, I think it warrants a comment in the code to make it clear to readers

File: chia/server/server.py [hunk 3]
```diff
@@ -419,14 +419,8 @@ async def start_client(
...
-
-            try:
-                if type(ip_address(target_node.host)) is IPv6Address:
-                    target_node = PeerInfo(f"[{target_node.host}]", target_node.port)
-            except ValueError:
-                pass
-
-            url = f"wss://{target_node.host}:{target_node.port}/ws"
+            ip = f"[{target_node.ip}]" if target_node.ip.is_v6 else f"{target_node.ip}"
+            url = f"wss://{ip}:{target_node.port}/ws"
```
Comment (arvidn) [inline chia/server/server.py:423]: maybe in the future it would make sense to make `PeerInfo` have a `__str__()` that contains this to-string logic

File: tests/simulation/test_simulator.py [hunk 4]
```diff
@@ -251,7 +251,7 @@ async def test_create_coins_with_invalid_amounts_raises(
 ) -> None:
     [[full_node_api], [[wallet_node, wallet_server]], _] = simulator_and_wallet
 
-    await wallet_server.start_client(PeerInfo("localhost", uint16(full_node_api.server._port)), None)
+    await wallet_server.start_client(PeerInfo("127.0.0.1", uint16(full_node_api.server._port)), None)
```
Comment (arvidn) [inline tests/simulation/test_simulator.py:254]: is this change necessary because we no longer support hostnames or is it a matter of reducing test flakiness?

**Review by arvidn — COMMENTED** [review]
looks good generally. Just a few questions

File: tests/simulation/test_simulator.py (hunk 4 above)
Comment (xdustinface) [reply tests/simulation/test_simulator.py:254]: The former, yes.

File: chia/server/address_manager.py (hunk 1 above)
Comment (xdustinface) [reply chia/server/address_manager.py:303]: It's just, before we validated the host in [PeerInfo.is_valid](https://github.com/Chia-Network/chia-blockchain/blob/99abc2d66f0136bf6d0420a9c3163664987a04ee/chia/types/peer_info.py#L16), see below. But with this PR we store valid IP addresses in `PeerInfo` so `is_valid` is not longer needed and gets dropped. The condition `addr.ip.is_private and not self.allow_private_subnets` is just the private subnet check which was nested into `PeerInfo.is_valid` for some reason.

File: chia/server/node_discovery.py (hunk 2 above)
Comment (xdustinface) [reply chia/server/node_discovery.py:68]: See 5009987591390d12aee466339e5d30e02709bcae.

**Review by arvidn — APPROVED** [review]
